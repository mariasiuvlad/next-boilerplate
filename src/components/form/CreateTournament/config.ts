import {
  Tournaments_Tournament_Status_Enum_Enum,
  Tournaments_Tournament_Type_Enum_Enum,
} from '__generated__/graphql'
import {InputConfig} from '../types'

/** @TODO move somewhere else */
const selectOptions = (sourceEnum: Record<string, string>) =>
  Object.values(sourceEnum).map((v) => ({
    value: v,
    label: v,
  }))

export const Title: InputConfig = {
  label: 'title',
  placeholder: 'title',
  type: 'text',
  name: 'title',
  id: 'title',
  bind: {
    required: 'Tournament title is required',
  },
}

export const Type = {
  label: 'type',
  placeholder: 'type',
  type: 'select',
  options: selectOptions(Tournaments_Tournament_Type_Enum_Enum),
  name: 'type',
  id: 'type',
  bind: {
    required: 'Tournament type is required',
  },
}

export const Status = {
  label: 'status',
  placeholder: 'status',
  type: 'select',
  options: selectOptions(Tournaments_Tournament_Status_Enum_Enum),
  name: 'status',
  id: 'status',
  bind: {
    required: 'Tournament status is required',
  },
}

export const StartDate = {
  label: 'start_date',
  placeholder: 'start_date',
  type: 'text',
  name: 'start_date',
  id: 'start_date',
  bind: {
    required: 'Tournament start_date is required',
  },
}

export const ParticipantsLimit = {
  label: 'participants_limit',
  placeholder: 'participants_limit',
  type: 'text',
  name: 'participants_limit',
  id: 'participants_limit',
  bind: {
    required: 'Tournament participants_limit is required',
  },
}

export const EntryFee = {
  label: 'entry_fee',
  placeholder: 'entry_fee',
  type: 'text',
  name: 'entry_fee',
  id: 'entry_fee',
  bind: {
    required: 'Tournament entry_fee is required',
  },
}
