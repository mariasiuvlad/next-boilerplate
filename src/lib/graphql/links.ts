import {HttpLink} from 'apollo-link-http'
import {WebSocketLink} from 'apollo-link-ws'
import {onError} from 'apollo-link-error'
import {ApolloLink, split} from 'apollo-link'
import {getMainDefinition} from 'apollo-utilities'
import {GQL_API} from '@config/index'
import {logger} from '@lib/log'

const log = logger.extend('gql')

const authLink = (jwtToken: string) =>
  new ApolloLink((operation, forward) => {
    if (jwtToken) {
      const headers = {Authorization: `Bearer ${jwtToken}`}
      operation.setContext({headers})
    }
    return forward(operation)
  })

const errorLink = onError(({graphQLErrors, networkError}) => {
  if (graphQLErrors)
    graphQLErrors.forEach(({message, locations, path}) =>
      log('[GraphQL error]: Message: %s, Location: %s, Path: %s', message, locations, path)
    )
  if (networkError) log('[Network error] %o', networkError)
})

const httpLink = new HttpLink({
  uri: GQL_API, // Server URL (must be absolute)
  credentials: 'same-origin', // Additional fetch() options like `credentials` or `headers`
})

// Create a WebSocket link
const wsLink = (jwtToken: string) =>
  // only create WebSocket link in the browser
  process.browser
    ? new WebSocketLink({
        uri: `ws://localhost:8080/v1/graphql`,
        options: {
          reconnect: true,
          connectionParams: {
            headers: {
              Authorization: `Bearer ${jwtToken}`,
            },
          },
        },
      })
    : null

const transportLink = (jwtToken: string) =>
  // only use WebSocket link in the browser
  process.browser
    ? split(
        // split based on operation type
        ({query}) => {
          const definition = getMainDefinition(query)
          return (
            definition.kind === 'OperationDefinition' && definition.operation === 'subscription'
          )
        },
        wsLink(jwtToken),
        httpLink
      )
    : httpLink

export const createClientLink = (jwtToken: string) => {
  return ApolloLink.from([errorLink, authLink(jwtToken), transportLink(jwtToken)])
}
