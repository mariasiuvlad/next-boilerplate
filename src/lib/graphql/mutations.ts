import gql from 'graphql-tag'

export const UPDATE_AVATAR = gql`
  mutation UpdateAvatar($avatar_url: String = "") {
    update_users(_set: {avatar_url: $avatar_url}, where: {}) {
      affected_rows
    }
  }
`

export const UPDATE_DISPLAY_NAME = gql`
  mutation UpdateDisplayName($display_name: String = "") {
    update_users(_set: {display_name: $display_name}, where: {}) {
      affected_rows
    }
  }
`

export const CREATE_TOURNAMENT = gql`
  mutation CreateTournament(
    $type: tournaments_tournament_type_enum_enum = elimination
    $title: String
    $status: tournaments_tournament_status_enum_enum = in_progress
    $start_date: date
    $participants_limit: Int = 10
    $entry_fee: float8
    $description: String
  ) {
    insert_tournaments_tournament_one(
      object: {
        description: $description
        entry_fee: $entry_fee
        participants_limit: $participants_limit
        start_date: $start_date
        status: $status
        title: $title
        type: $type
      }
    ) {
      id
    }
  }
`

export const JoinTournament = gql`
  mutation JoinTournament($tournament_id: uuid) {
    insert_tournaments_tournament_participants_one(object: {tournament_id: $tournament_id}) {
      user_id
      tournament_id
    }
  }
`

export const LeaveTournament = gql`
  mutation LeaveTournament(
    $tournament_id: uuid_comparison_exp = {}
    $user_id: uuid_comparison_exp = {}
  ) {
    delete_tournaments_tournament_participants(
      where: {_and: {tournament_id: $tournament_id, user_id: $user_id}}
    ) {
      affected_rows
    }
  }
`
