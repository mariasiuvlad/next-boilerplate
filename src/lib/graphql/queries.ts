import gql from 'graphql-tag'

export const GET_USER = gql`
  query GetUser {
    me {
      id
      display_name
      avatar_url
      updated_at
      created_at
    }
  }
`

export const GetTournaments = gql`
  fragment TournamentInfo on tournaments_tournament {
    id
    owner {
      display_name
      avatar_url
    }
    participants {
      user_id
      user {
        display_name
        avatar_url
      }
    }
    entry_fee
    description
    participants_limit
    start_date
    status
    title
    type
  }

  query GetTournaments($limit: Int = 10, $offset: Int = 0) {
    tournaments_tournament(limit: $limit, offset: $offset) {
      ...TournamentInfo
    }
  }

  query GetTournament($id: uuid!) {
    tournaments_tournament_by_pk(id: $id) {
      ...TournamentInfo
    }
  }
`

export const CountTournaments = gql`
  query CountTournaments {
    tournaments_tournament_aggregate {
      aggregate {
        count
      }
    }
  }
`
