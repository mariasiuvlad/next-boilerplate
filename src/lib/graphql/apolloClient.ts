import {useMemo} from 'react'
import {ApolloClient} from 'apollo-client'
import {InMemoryCache, NormalizedCacheObject} from 'apollo-cache-inmemory'
import {createClientLink} from './links'

let apolloClient: ApolloClient<NormalizedCacheObject>

type ApolloInitOptions = {
  initialState?: any
  jwtToken?: string
}

function createApolloClient(jwtToken = '') {
  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: createClientLink(jwtToken),
    cache: new InMemoryCache(),
  })
}

export function initializeApollo({initialState, jwtToken}: ApolloInitOptions) {
  const _apolloClient = jwtToken
    ? createApolloClient(jwtToken)
    : apolloClient ?? createApolloClient()

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    _apolloClient.cache.restore(initialState)
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === 'undefined') return _apolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient

  return _apolloClient
}

export function useApollo(initialState = {}, jwtToken = '') {
  const store = useMemo(() => initializeApollo({initialState, jwtToken}), [
    jwtToken,
  ])
  return store
}
