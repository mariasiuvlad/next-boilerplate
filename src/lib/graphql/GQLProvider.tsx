import {ApolloProvider} from '@apollo/react-hooks'
import {useApollo} from './apolloClient'
import {useAuth} from '@context/auth'

const GQLProvider = (props) => {
  const {initialApolloState, children} = props
  const {data} = useAuth()
  const client = useApollo(initialApolloState, data?.jwt_token)
  return <ApolloProvider client={client}>{children}</ApolloProvider>
}

export default GQLProvider
