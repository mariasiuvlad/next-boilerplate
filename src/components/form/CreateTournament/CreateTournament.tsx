import Button from '@components/common/Button'
import {ErrorMessage} from '@hookform/error-message'
import FormError from '../components/FormError'
import {CreateTournamentMutationVariables} from '__generated__/graphql'
import {UseFormMethods} from 'react-hook-form'
import {UseInput} from '../types'

interface CreateTournamentProps {
  inputs: UseInput[]
  onSubmit(): void
  form: UseFormMethods<CreateTournamentMutationVariables>
}

export default function CreateTournament({
  inputs,
  onSubmit,
  form: {
    errors,
    formState: {isSubmitting},
  },
}: CreateTournamentProps) {
  return (
    <form onSubmit={onSubmit}>
      <section className="mb-4 mt-16">
        <ErrorMessage errors={errors} name="api" render={FormError} />
      </section>
      <section className="mb-8">
        <h1 className="text-yellow-500 text-2xl text-center">Create Tournament</h1>
        {inputs.map((input) => (
          <div className="my-4" key={input.name}>
            {input.Control}
          </div>
        ))}
      </section>
      <Button disabled={isSubmitting} variant="primary" label="Create Tournament" />
    </form>
  )
}
