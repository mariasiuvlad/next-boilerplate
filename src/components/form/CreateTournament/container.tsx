import {useCreateTournamentMutation, CreateTournamentMutationVariables} from '__generated__/graphql'
import {useCallback} from 'react'
import CreateTournament from './CreateTournament'
import {useForm} from 'react-hook-form'
import useInput from '../components/useInput'
import * as Config from './config'

/** @TODO turn into form container template */
export default function CreateTournamentContainer() {
  const form = useForm<CreateTournamentMutationVariables>()
  const {handleSubmit, reset, setError} = form

  const [createTournament] = useCreateTournamentMutation()

  const onSubmit = useCallback(async (variables: CreateTournamentMutationVariables) => {
    createTournament({variables}).then(
      () => reset(),
      ({message}) => setError('api', {type: 'manual', message})
    )
  }, [])

  const textInput = useInput(form)
  return (
    <CreateTournament
      inputs={Object.values(Config).map(textInput)}
      form={form}
      onSubmit={handleSubmit(onSubmit)}
    />
  )
}
