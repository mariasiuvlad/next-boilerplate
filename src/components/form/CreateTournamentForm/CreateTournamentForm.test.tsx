import CreateTournament from './container'
import {render, fireEvent, screen, waitFor} from '@testing-library/react'
import {MockedProvider} from '@apollo/react-testing'
import {CreateTournamentDocument, GetTournamentsDocument} from '__generated__/graphql'

let mutationCalled = false

export const CreateTournamentMock = {
  request: {
    query: CreateTournamentDocument,
    variables: {
      type: 'elimination',
      status: 'in_progress',
      participants_limit: '2',
      title: 'test title',
      start_date: '10/10/2020',
      entry_fee: '100',
    },
  },
  result: () => {
    mutationCalled = true
    return {data: {affected_rows: 1}}
  },
}

export const GetTournamentsMock = {
  request: {query: GetTournamentsDocument, variables: {limit: 10, offset: 0}},
  result: () => ({data: {}}),
}

const submit = () => fireEvent.submit(screen.getByRole('button'))
const setTitle = (value) =>
  fireEvent.input(screen.getByLabelText('title'), {
    target: {value},
  })

const setStartDate = (value) =>
  fireEvent.input(screen.getByLabelText('start_date'), {
    target: {value},
  })

const setParticipantsLimit = (value) =>
  fireEvent.input(screen.getByLabelText('participants_limit'), {
    target: {value},
  })

const setEntryFee = (value) =>
  fireEvent.input(screen.getByLabelText('entry_fee'), {
    target: {value},
  })

test('it renders', () => {
  const tree = render(
    <MockedProvider mocks={[]}>
      <CreateTournament />
    </MockedProvider>
  )
  expect(tree).toMatchSnapshot()
})

test('it shows errors', async () => {
  const {getAllByRole} = render(
    <MockedProvider mocks={[CreateTournamentMock, GetTournamentsMock]}>
      <CreateTournament />
    </MockedProvider>
  )
  submit()
  await waitFor(() => expect(getAllByRole('alert')).not.toHaveLength(0))
})

test('it submits', async () => {
  const {queryAllByRole} = render(
    <MockedProvider mocks={[CreateTournamentMock, GetTournamentsMock]}>
      <CreateTournament />
    </MockedProvider>
  )
  setTitle('test title')
  setStartDate('10/10/2020')
  setParticipantsLimit(2)
  setEntryFee(100)
  submit()
  await waitFor(() => expect(queryAllByRole('alert')).toHaveLength(0))
  await waitFor(() => expect(mutationCalled).toBe(true))
})
