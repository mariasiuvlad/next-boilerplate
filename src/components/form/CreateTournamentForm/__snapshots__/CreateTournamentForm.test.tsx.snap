// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <section
          class="mb-4 mt-16"
        />
        <section
          class="mb-8"
        >
          <h1
            class="text-yellow-500 text-2xl text-center"
          >
            Create Tournament
          </h1>
          <div
            class="my-4"
          >
            <label
              class="label"
              for="title"
            >
              title
            </label>
            <div
              class="container"
            >
              <input
                class="input"
                id="title"
                name="title"
                placeholder="title"
                type="text"
              />
            </div>
          </div>
          <div
            class="my-4"
          >
            <label
              class="label"
              for="type"
            >
              type
            </label>
            <select
              class="select"
              id="type"
              name="type"
            >
              <option
                class="p-8"
                value="elimination"
              >
                elimination
              </option>
              <option
                class="p-8"
                value="group"
              >
                group
              </option>
            </select>
          </div>
          <div
            class="my-4"
          >
            <label
              class="label"
              for="status"
            >
              status
            </label>
            <select
              class="select"
              id="status"
              name="status"
            >
              <option
                class="p-8"
                value="in_progress"
              >
                in_progress
              </option>
            </select>
          </div>
          <div
            class="my-4"
          >
            <label
              class="label"
              for="start_date"
            >
              start_date
            </label>
            <div
              class="container"
            >
              <input
                class="input"
                id="start_date"
                name="start_date"
                placeholder="start_date"
                type="text"
              />
            </div>
          </div>
          <div
            class="my-4"
          >
            <label
              class="label"
              for="participants_limit"
            >
              participants_limit
            </label>
            <div
              class="container"
            >
              <input
                class="input"
                id="participants_limit"
                name="participants_limit"
                placeholder="participants_limit"
                type="text"
              />
            </div>
          </div>
          <div
            class="my-4"
          >
            <label
              class="label"
              for="entry_fee"
            >
              entry_fee
            </label>
            <div
              class="container"
            >
              <input
                class="input"
                id="entry_fee"
                name="entry_fee"
                placeholder="entry_fee"
                type="text"
              />
            </div>
          </div>
        </section>
        <button
          class="button primary"
        >
          <span
            class="label"
          >
            Create Tournament
          </span>
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <section
        class="mb-4 mt-16"
      />
      <section
        class="mb-8"
      >
        <h1
          class="text-yellow-500 text-2xl text-center"
        >
          Create Tournament
        </h1>
        <div
          class="my-4"
        >
          <label
            class="label"
            for="title"
          >
            title
          </label>
          <div
            class="container"
          >
            <input
              class="input"
              id="title"
              name="title"
              placeholder="title"
              type="text"
            />
          </div>
        </div>
        <div
          class="my-4"
        >
          <label
            class="label"
            for="type"
          >
            type
          </label>
          <select
            class="select"
            id="type"
            name="type"
          >
            <option
              class="p-8"
              value="elimination"
            >
              elimination
            </option>
            <option
              class="p-8"
              value="group"
            >
              group
            </option>
          </select>
        </div>
        <div
          class="my-4"
        >
          <label
            class="label"
            for="status"
          >
            status
          </label>
          <select
            class="select"
            id="status"
            name="status"
          >
            <option
              class="p-8"
              value="in_progress"
            >
              in_progress
            </option>
          </select>
        </div>
        <div
          class="my-4"
        >
          <label
            class="label"
            for="start_date"
          >
            start_date
          </label>
          <div
            class="container"
          >
            <input
              class="input"
              id="start_date"
              name="start_date"
              placeholder="start_date"
              type="text"
            />
          </div>
        </div>
        <div
          class="my-4"
        >
          <label
            class="label"
            for="participants_limit"
          >
            participants_limit
          </label>
          <div
            class="container"
          >
            <input
              class="input"
              id="participants_limit"
              name="participants_limit"
              placeholder="participants_limit"
              type="text"
            />
          </div>
        </div>
        <div
          class="my-4"
        >
          <label
            class="label"
            for="entry_fee"
          >
            entry_fee
          </label>
          <div
            class="container"
          >
            <input
              class="input"
              id="entry_fee"
              name="entry_fee"
              placeholder="entry_fee"
              type="text"
            />
          </div>
        </div>
      </section>
      <button
        class="button primary"
      >
        <span
          class="label"
        >
          Create Tournament
        </span>
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
